src/libostree/ostree-repo-finder.h

```diff
@@ -0,0 +1,116 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ * Authors:
+ *  - Philip Withnall <withnall@endlessm.com>
+ */
+
+#pragma once
+
+#include <gio/gio.h>
+#include <glib.h>
+#include <glib-object.h>
+
+#include "ostree-types.h"
+
+G_BEGIN_DECLS
+
+#define OSTREE_TYPE_REPO_FINDER (ostree_repo_finder_get_type ())
+
+_OSTREE_PUBLIC
+G_DECLARE_INTERFACE (OstreeRepoFinder, ostree_repo_finder, OSTREE, REPO_FINDER, GObject)
+
+struct _OstreeRepoFinderInterface
+{
+  GTypeInterface g_iface;
+
+  void (*resolve_async) (OstreeRepoFinder    *self,
+                         const gchar * const *refs,
+                         GCancellable        *cancellable,
+                         GAsyncReadyCallback  callback,
+                         gpointer             user_data);
+  GPtrArray *(*resolve_finish) (OstreeRepoFinder  *self,
+                                GAsyncResult      *result,
+                                GError           **error);
+};
+
+_OSTREE_PUBLIC
+void ostree_repo_finder_resolve_async (OstreeRepoFinder    *self,
+                                       const gchar * const *refs,
+                                       GCancellable        *cancellable,
+                                       GAsyncReadyCallback  callback,
+                                       gpointer             user_data);
+_OSTREE_PUBLIC
+GPtrArray *ostree_repo_finder_resolve_finish (OstreeRepoFinder  *self,
+                                              GAsyncResult      *result,
+                                              GError           **error);
+
+_OSTREE_PUBLIC
+void ostree_repo_finder_resolve_all_async (OstreeRepoFinder * const *finders,
+                                           const gchar * const      *refs,
+                                           GCancellable             *cancellable,
+                                           GAsyncReadyCallback       callback,
+                                           gpointer                  user_data);
+_OSTREE_PUBLIC
+GPtrArray *ostree_repo_finder_resolve_all_finish (GAsyncResult  *result,
+                                                  GError       **error);
+
+/**
+ * OstreeRepoFinderResult:
+ * @uri: TODO
+ * @priority: TODO
+ *
+ * TODO: basically a structure which says ‘you can download these refs from
+ * this URI, and they will be up to date’, along with some relevant metadata.
+ *
+ * Since: 2017.6
+ */
+typedef struct
+{
+  /* TODO: Should I extend this to also support mirrors and metalink? Maybe add a GVariant *metadata too?
+   * For example, should a result be able to force GPG verification? */
+  /* TODO: would it support torrenting? */
+  gchar *uri;
```

Comment:
> > Synthesizing a remote internally would be good. Right now it's remotes that specify GPG for example (and in some cases which key). So perhaps this API should take a remote as input and use its GPG config?
> 
> I think we’re agreed on synthesising a remote and doing `s/gchar *uri/gchar *remote_name/` then. 👍

Gah, this won’t work, since the remotes are cached in the `OstreeRepo`, which the `OstreeRepoFinder` implementations don’t have access to (and it doesn’t make sense for them to have access to). I’ll see what I can come up with, but it might mean making `OstreeRemote` public (even if most API continues to use `const gchar *remote` to access remotes by name, for simplicity).